tempesta_fw/cache.c

```diff
@@ -1743,10 +1742,22 @@ static void
 tfw_wq_tasklet(unsigned long data)
 {
 	TfwWorkTasklet *ct = (TfwWorkTasklet *)data;
+	TfwRBQueue *wq = &ct->wq;
 	TfwCWork cw;
 
-	while (!tfw_wq_pop(&ct->wq, &cw))
+	while (!tfw_wq_pop(wq, &cw))
 		tfw_cache_do_action(cw.msg, cw.action);
+
+	set_bit(TFW_QUEUE_B_IPI, &wq->flags);
+
+	smp_mb__after_atomic();
```

Comment:
OK, here we set the IPI flag before checking the queue size and we check queue size before the flag in `tfw_wq_push()` - the synchronization seems safe. We need the barrier to make the queue size and the flag dependant. However, it seems the [barrier pairing rule](https://github.com/tempesta-tech/linux-4.14.32-tfw/blob/master/Documentation/memory-barriers.txt#L932) is broken here: strictly speaking, there is no guarantee that the queue operation won't be reordered with the flag testing in `tfw_wq_push()`. I see that `__tfw_wq_push()` uses at least atomic set, but it seems we still need an `_after_atomic()` barrier in `tfw_wq_push()` now.

Reply:
Corrected.